op-e2e/actions/proofs/garbage_channel_test.go

```diff
@@ -0,0 +1,144 @@
+		env.miner.ActL1StartBlock(12)(t)
+		env.miner.ActL1IncludeTxByHash(env.batcher.LastSubmitted.Hash())(t)
+		env.miner.ActL1EndBlock(t)
+
+		// Finalize the block with the first channel frame on L1.
+		env.miner.ActL1SafeNext(t)
+		env.miner.ActL1FinalizeNext(t)
+
+		// Instruct the sequencer to derive the L2 chain from the data on L1 that the batcher just posted.
+		env.sequencer.ActL1HeadSignal(t)
+		env.sequencer.ActL2PipelineFull(t)
+	}
+
+	const NumL2Blocks = 10
+
+	// Build NumL2Blocks empty blocks on L2
+	for i := 0; i < NumL2Blocks; i++ {
+		env.sequencer.ActL2StartBlock(t)
+		env.sequencer.ActL2EndBlock(t)
+	}
+
+	// Buffer the first half of L2 blocks in the batcher, and submit it.
+	for i := 0; i < NumL2Blocks/2; i++ {
+		env.batcher.ActL2BatchBuffer(t)
+	}
+	env.batcher.ActL2BatchSubmit(t)
+
+	// Include the batcher transaction.
+	includeBatchTx(env)
+
+	// Ensure that the safe head has not advanced - the channel is incomplete.
+	l2SafeHead := env.engine.L2Chain().CurrentSafeBlock()
+	require.Equal(t, uint64(0), l2SafeHead.Number.Uint64())
+
+	// Buffer the second half of L2 blocks in the batcher.
+	for i := 0; i < NumL2Blocks/2; i++ {
+		env.batcher.ActL2BatchBuffer(t)
+	}
+	env.batcher.ActL2ChannelClose(t)
+	expectedSecondFrame := env.batcher.ReadNextOutputFrame(t)
+
+	// Submit a garbage frame, modified from the expected second frame.
+	env.batcher.ActL2BatchSubmitGarbageRaw(t, expectedSecondFrame, garbageKind)
+	// Include the garbage second frame tx
+	includeBatchTx(env)
+
+	// Ensure that the safe head has not advanced - the channel is incomplete.
+	l2SafeHead = env.engine.L2Chain().CurrentSafeBlock()
+	require.Equal(t, uint64(0), l2SafeHead.Number.Uint64())
+
+	// Submit the correct second frame.
+	env.batcher.ActL2BatchSubmitRaw(t, expectedSecondFrame)
+	// Include the corract second frame tx.
+	includeBatchTx(env)
+
+	// Ensure that the safe head has advanced - the channel is complete.
+	l2SafeHead = env.engine.L2Chain().CurrentSafeBlock()
+	require.Equal(t, uint64(NumL2Blocks), l2SafeHead.Number.Uint64())
+
+	// Run the FPP on L2 block # NumL2Blocks. The claim is honest, so it should pass.
+	err := env.RunFaultProofProgram(t, gt, NumL2Blocks, inputParams...)
+	checkResult(gt, err)
+}
+
+func Test_ProgramAction_GarbageChannel_HonestClaim_Granite(gt *testing.T) {
+	for _, garbageKind := range garbageKinds {
+		gt.Run(garbageKind.String(), func(t *testing.T) {
+			runGarbageChannelTest(
+				t,
+				garbageKind,
+				func(gt *testing.T, err error) {
+					require.NoError(gt, err, "fault proof program should not have failed")
+				},
+			)
+		})
+	}
+}
+
+func Test_ProgramAction_GarbageChannel_JunkClaim_Granite(gt *testing.T) {
+	for _, garbageKind := range garbageKinds {
+		gt.Run(garbageKind.String(), func(t *testing.T) {
+			runGarbageChannelTest(
+				t,
+				garbageKind,
+				func(gt *testing.T, err error) {
+					require.ErrorIs(gt, err, claim.ErrClaimNotValid, "fault proof program should have failed")
```

Comment:
Probably worth having a helper method that can assert valid or invalid results from the fault proof program  - maybe just pass the expected result to `env.RunFaultProofProgram`?

Reply:
For sure! Can follow up in another PR. I think we should probably keep the functional check, but maybe something like `WithResult` as a short-hand when we know what error we're getting back beforehand?

Reply:
Yeah I'd be happy with that. I'm not sure if there's really a time we want to execute the program and not know if it should be valid or not but I think a `WithResult` type thing makes the API clearer than a bool anyway.

And yeah happy for it to be a follow up PR.